### PR TITLE
[WIP][AIRFLOW-594] Add dag source plugin hooks

### PR DESCRIPTION
**Pull request by drewsonne**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-594\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-594
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-594\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

  This PR is a precursor to loading DAGs from pip packages. It provides a new plugin structure, `AirflowPluginDagSource` with two methods `collect_dags_from_plugins` and `put_dag_plugins_on_disk` where a Plugin author can:
 -  implement the former and provide logic to inject the `airflow.model.DAG` objects into a provided `airflow.modal.DagBag` object using their own custom logic,
 - or implement the latter and provide logic to write the dag python files to disk in the dag folder where they will be collected by the normal DAG loading setup.

From discussion below, potential use cases could be:
 - to load dags from pip packages
 - to load dags from (for example) S3, GCP Cloud Storage, Azure Blob Storage (using the built in hooks/connections);
 - integration with say, https://github.com/wooga/airconditioner without mixing yaml and python for dag definitions;
 - tighter integration with https://github.com/python-bonobo/bonobo
 - as an intermediary step to handle some automatic lineage injection on DAG load
 - perform some validation or linting of DAGs before loading.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by bjoernpollex-sc** [comment]
It feels like creating a plugin to deploy DAGs is an unnecessary extra step. Also, DAGs aren't really plugins, right? Wouldn't it be easier to just provide a separate entry-point for DAGs? #730, which never got merged, has an example of this.

Btw., thanks for working on this, and also #4412, these are great improvements!

**Comment by drewsonne** [comment]
This doesn't assume DAGs are plugins, and although it would be easier for this specific functionality, yes, if someone wanted to:
 - to load dags from (for example) S3, GCP Cloud Storage, Azure Blob Storage (using the built in hooks);
 - integration with say, https://github.com/wooga/airconditioner without mixing yaml *and* python for dag definitions; 
 - tighter integration with https://github.com/python-bonobo/bonobo
 - or as an intermediary step to handle some automatic lineage injection on DAG load

then we need new PRs for each of those pieces of functionality and wait for the release cycle, unless we're building airflow from src in our prod env's.

If this (eventually, after I clean the PR up) is approved, the next step is a built-in plugin to load DAGs from entrypoints.

**Comment by bjoernpollex-sc** [comment]
Great points, I was too focused on my own particular use-case!

**Comment by drewsonne** [comment]
You did make me think some of these through a bit harder, and I think I should change the PR to reflect :-)

File: docs/conf.py [hunk 1]
```diff
@@ -122,7 +122,10 @@
     'docroles'
 ]
 
-autodoc_default_flags = ['show-inheritance', 'members']
+autodoc_default_options = {
```
Comment (mik-laj) [inline docs/conf.py]: This look like a unreleated changes. This should be done in a separate PR.

File: docs/plugins.rst [hunk 2]
```diff
@@ -20,7 +20,8 @@ Plugins
 
 Airflow has a simple plugin manager built-in that can integrate external
 features to its core by simply dropping files in your
-``$AIRFLOW_HOME/plugins`` folder.
+``$AIRFLOW_HOME/plugins`` folder or providing pip packages which provide
```
Comment (mik-laj) [inline docs/plugins.rst:23]: Good point. I thought about it yesterday and I was planning to improve it.

File: docs/plugins.rst [hunk 3]
```diff
@@ -127,6 +130,9 @@ For example,
 
 * For ``Operator`` plugin, an ``execute`` method is compulsory.
 * For ``Sensor`` plugin, a ``poke`` method returning a Boolean value is compulsory.
+* For ``AirflowPluginDagSource`` plugin, ``name``, ``add_dags_to_dagbag``, and
```
Comment (mik-laj) [inline docs/plugins.rst:133]:
```suggestion
* For ::class:`~airflow.plugin.AirflowPluginDagSource` plugin, ``name``, ``add_dags_to_dagbag``, and
```
This will allow automatic linking to the class documentation when such documentation will be created.

File: docs/conf.py (hunk 1 above)
Comment (drewsonne) [reply docs/conf.py]: yeah, I was trying to figure out why the doc step was failing in the build on travis

File: docs/plugins.rst (hunk 3 above)
Comment (drewsonne) [reply docs/plugins.rst:133]: I gotta fix the test cases and then I'll crack on with the doc problems :-D
